pageserver/src/virtual_file.rs

```diff
@@ -1206,13 +1179,16 @@ mod tests {
     impl MaybeVirtualFile {
         async fn read_exact_at(
             &self,
-            mut buf: Vec<u8>,
+            mut slice: tokio_epoll_uring::Slice<Vec<u8>>,
             offset: u64,
             ctx: &RequestContext,
-        ) -> Result<Vec<u8>, Error> {
+        ) -> Result<tokio_epoll_uring::Slice<Vec<u8>>, Error> {
             match self {
-                MaybeVirtualFile::VirtualFile(file) => file.read_exact_at(buf, offset, ctx).await,
-                MaybeVirtualFile::File(file) => file.read_exact_at(&mut buf, offset).map(|()| buf),
+                MaybeVirtualFile::VirtualFile(file) => file.read_exact_at(slice, offset, ctx).await,
+                MaybeVirtualFile::File(file) => {
+                    let rust_slice: &mut [u8] = slice.as_mut_rust_slice_full_zeroed();
```

Comment:
This makes sense, but it's a bit odd that only the first caller from `VirtualFile::read_exact_at_impl` will make `as_mut_rust_slice_full_zeroed` will do anything. Quite nice having the rust slice initialization here though

Reply:
>  odd that only the first caller from VirtualFile::read_exact_at_impl will make as_mut_rust_slice_full_zeroed will do anything. 

Yeah, typically `as_` APIs are supposed to be cheap.

If more of the MaybeUninit API were stabilitized, perhaps this would have made sense:

```rust
fn as_mut_rust_slice_full() -> MaybeUninit<&mut [u8]>
```

Then the caller could call [MaybeUninit::fill](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.fill)

But, I don't think `MaybeUninit<&mut [u8]>` exists even, so, something is missing.